### PR TITLE
network: socket interface support for addresses

### PR DESCRIPTION
**Pull request by florincoras**
Signed-off-by: Florin Coras <fcoras@cisco.com>

Risk Level: Low
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by florincoras** [comment]
cc @yanavlasov @htuch Is this in combination with the support for configurable `SocketInterface` enough to solve #11618?

File: source/common/network/address_impl.h [hunk 1]
```diff
@@ -54,23 +60,23 @@ class Ipv4Instance : public InstanceBase {
   /**
    * Construct from an existing unix IPv4 socket address (IP v4 address and port).
    */
-  explicit Ipv4Instance(const sockaddr_in* address);
+  explicit Ipv4Instance(const sockaddr_in* address, const std::string& sock_interface = "");
```
Comment (htuch) [inline source/common/network/address_impl.h:63]: Nit: `absl::string_view` for all `const std::string&`s.

File: source/common/network/socket_impl.cc [hunk 2]
```diff
@@ -11,12 +11,24 @@
...
 namespace Network {
 
+static IoHandlePtr ioHandleForAddr(Socket::Type type, const Address::InstanceConstSharedPtr addr) {
+  auto sock_interface_name = addr->socketInterface();
+  if (!sock_interface_name.empty()) {
+    auto sock_interface = const_cast<SocketInterface*>(socketInterface(sock_interface_name));
+    if (sock_interface != nullptr) {
+      return sock_interface->socket(type, addr);
+    }
+    return nullptr;
```
Comment (yanavlasov) [inline source/common/network/socket_impl.cc:21]: This will cause a SIGSEGV sometime later, right? Maybe the right thing to do is to RELEASE_ASSERT here that the factory for creating IO handles is missing.

**Comment by yanavlasov** [comment]
Also would you be able to make unit test exercising this, please?

File: source/common/network/address_impl.h (hunk 1 above)
Comment (florincoras) [reply source/common/network/address_impl.h]: Done!

File: source/common/network/socket_impl.cc (hunk 2 above)
Comment (florincoras) [reply source/common/network/socket_impl.cc]: Fair point! Made it a RELEASE_ASSERT.

**Comment by florincoras** [comment]
> Also would you be able to make unit test exercising this, please?

Done! Added an integration test which opens a connection with an address that explicitly asks for the default socket interface

**Comment by florincoras** [comment]
@yanavlasov, I suspect the arm failure is just a flake. After you get a chance to review the PR, I'll merge master and hopefully it'll go away.

**Comment by yanavlasov** [comment]
@florincoras thanks it looks good. Would it be possible to also have a test that tests complete plumbing with a custom `SocketInterface`, bootstrap extension, etc. This will also server as a documentation for anyone who needs to create their `SocketInterface`.
Super excited about this one. I could use it in the integration tests where the test needs to simulate TCP back-pressure from the peer.

**Comment by florincoras** [comment]
> @florincoras thanks it looks good. Would it be possible to also have a test that tests complete plumbing with a custom `SocketInterface`, bootstrap extension, etc. This will also server as a documentation for anyone who needs to create their `SocketInterface`.

I suspect my description above was misleading. The default `SocketInterface` implemented [here](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_interface_impl.cc) actually goes through the whole process of [registering a factory](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_interface_impl.cc#L101) which can be configured through a bootstrap extension (albeit no actual config params are supported currently) and subsequently can be looked up by name via [this](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_interface.h#L46). So, the integration test actually exercises both [bootstrap extensions](https://github.com/envoyproxy/envoy/blob/master/test/integration/socket_interface_integration_test.cc#L22) and the `SocketInterface` lookup by [name](https://github.com/envoyproxy/envoy/pull/12189/files#diff-d28ab6f8cef00963ff99f85579f7308dR65). 

In short, to implement a new, configurable `SocketInterface`, apart from the actual socket apis, one only needs implementations for [these](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_interface_impl.h#L20-L27). 

Also, [this](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_interface_impl.cc#L103-L104) static global init is why we don't always need to configure a default interface in the startup config file. 

> Super excited about this one. I could use it in the integration tests where the test needs to simulate TCP back-pressure from the peer.

Cool idea! :-D

File: source/common/network/socket_interface.h [hunk 3]
```diff
@@ -52,5 +52,23 @@ static inline const SocketInterface* socketInterface(std::string name) {
...
+ * Create IoHandle for given address
+ * @param type type of socket to be requested
+ * @param addr address that is gleaned for address type, version and socket interface name
+ * @return @ref Network::IoHandlePtr that wraps the underlying socket file descriptor
+ */
+static inline IoHandlePtr ioHandleForAddr(Socket::Type type,
+                                          const Address::InstanceConstSharedPtr addr) {
+  auto sock_interface_name = addr->socketInterface();
+  if (!sock_interface_name.empty()) {
+    auto sock_interface = const_cast<SocketInterface*>(socketInterface(sock_interface_name));
```
Comment (mattklein123) [inline source/common/network/socket_interface.h:65]: nit: this const_cast seems a bit off. Any reason to not just fix the socket method itself to be const?

Comment (florincoras) [reply source/common/network/socket_interface.h:65]: I thought about that too but `socket()` could end up changing the `SocketInterface` object. The alternative would be to remove const from `SocketInterface` and trust that callers won't do anything irresponsible.

Comment (mattklein123) [reply source/common/network/socket_interface.h:65]: Why would it change the interface object? Does that happen in practice right now? Can we cross that bridge later if we come to it? If it really is a mutable thing I would just remove const.

Comment (florincoras) [reply source/common/network/socket_interface.h:65]: I was thinking that interface objects could be tracking the sockets, e.g., they could be building some form of socket descriptor map. We're not doing that now, as we're offloading everything to the kernel, but I did hit this particular issue with a custom socket interface implementation. But okay, let's cross that bridge later :-)

File: source/common/network/socket_interface.h [hunk 4]
```diff
@@ -52,5 +52,23 @@ static inline const SocketInterface* socketInterface(std::string name) {
...
+ * @param type type of socket to be requested
+ * @param addr address that is gleaned for address type, version and socket interface name
+ * @return @ref Network::IoHandlePtr that wraps the underlying socket file descriptor
+ */
+static inline IoHandlePtr ioHandleForAddr(Socket::Type type,
+                                          const Address::InstanceConstSharedPtr addr) {
+  auto sock_interface_name = addr->socketInterface();
+  if (!sock_interface_name.empty()) {
+    auto sock_interface = socketInterface(sock_interface_name);
+    RELEASE_ASSERT(sock_interface != nullptr,
```
Comment (antoniovicente) [inline source/common/network/socket_interface.h:66]: Regarding this RELEASE_ASSERT, what guarantees that the interface names returned by the address are valid?

Comment (florincoras) [reply source/common/network/socket_interface.h:66]: By default, the names are null. If anybody sets a name on the address, e.g., resolvers, the expectation is that those names are valid. The alternative would be to silently accept errors and use the default `SocketInterface` implementation but that would just hide misconfiguration. Still, we can change it if need be!

Comment (antoniovicente) [reply source/common/network/socket_interface.h:66]: We need config validation so this doesn't turn into a runtime failure at some indeterminate time after a configuration that refers to an unknown socket interface provider is specified.

Crashing the proxy in a situation like this doesn't seem like the right thing.  Terminating the connection with some critical errors may be an option, but really - we need some protections during config load to provide some better guarantees that this won't happen.

Comment (florincoras) [reply source/common/network/socket_interface.h:66]: Currently, there's no way to configure a socket interface name on an address via config. The addresses have resolvers and resolvers spawn `Address::Instance`s, with or without a socket interface name. So then, if the resolvers are misconfigured and generate broken addresses, what should we do? 

But, let's assume we'll allow configuration of addresses with socket interface names and don't validate those in resolvers or wherever we spawn `Address::Instance`s. Then, we could return nullptr `IoHandlePtr`s or an `IoSocketHandleImpl` with a -1 fd. Consequently, any place that constructs a `Socket` must afterwards check if the internal `IoHandle` was correctly initialized. A quick code check shows that `ConnectionImpl` `RELEASE_ASSERT`s that the socket it was given is valid in the constructor and similarly `NetworkListenSocket` `RELEASE_ASSERT`s that the `IoHandle`'s fd is valid.  

Do we any other alternatives?

Comment (florincoras) [reply source/common/network/socket_interface.h:66]: That's an interesting idea! It's definitely doable, assuming all the places that construct `Address::Instance`s have access to the `socket_interface.h` and therefore ability to lookup the `SocketInterface` implementations. But I'll defer this to @yanavlasov as I'm sure he has more experience with this type of problems given #11618
